doc/developer/design/20241028_optimizer_releases.md

```diff
@@ -0,0 +1,81 @@
+# Optimizer Release Engineering
+
+## The Problem
+
+Currently, users run on "Materialize weekly", but:
+
+  1. That cannot be case forever, as self-hosted users may not upgrade immediately.
+  2. That is not always good for users, who can experience regressions when optimizer changes don't suit their workloads.
+  3. That is not always good for people working on the optimizer, who must size their work and gate it with feature flags to accommodate the weekly release schedule.
+
+## Success Criteria
+
+Customers---self-hosted or cloud---will be able to qualify new optimizers before migrating to them.
+
+Optimizer engineers will be able to develop features with confidence, namely:
+
+  - introducing new transforms
+  - updating existing transforms
+  - targeting new dataflow operators
+  - changing AST types for HIR, MIR, or LIR
```

Comment:
I agree that (2) is the best possible thing, and it seems unavoidable. I just can't imagine this company without having that at some point in the next 1-2 years. Do you maybe have a ballpark estimate for how much effort will it be to develop (2)? I guess it would be an evolution of the 0-dt machinery, so that it could be run outside an actual release rollout, right?

We'll probably also do some amount of (1), because running synthetic benchmarks is more lightweight than either (2) or (3), so it will be great to run it on PRs. But I can't really imagine (1) ever being comprehensive enough to catch ~all user regressions. As we'll have more and more users, they'll just keep running into new edge cases.

(3) is ok only as a temporary measure before we develop (2). It would allow us to at least partially shift some blame to users for outages caused by "sudden" regressions, when they had several months to test a new version but didn't do it. But this is far from ideal.

(Btw. if we have (2), then we could also consider having only two parallel optimizer versions instead of three: An unstable, and just one stable (instead of two stables). This is because we could catch ~all regressions before promoting an unstable to stable. Still, the drawback of having only two parallel versions would be that sometimes the unstable-to-stable promotion would be delayed by possibly many weeks due to addressing some regression, in which time all the other users for whom there is no regression wouldn't be able to get the benefits of the new version. Another drawback of having just two versions would be that normal optimizer work would need to be put on hold during that period when we are fixing regressions just before an unstable-to-stable promotion, to avoid introducing more regressions.)

Reply:
The core issue---the thing that prompted discussions of release engineering---is that we're not confident qualifying optimizer releases. There's planned work to improve the situation there, and it's possible that's all we need.

Reply:
> I guess it would be an evolution of the 0-dt machinery, so that it could be run outside an actual release rollout, right?

Yeah, exactly—much of the infrastructure already exists thanks to 0dt. @teskje has a "shadow replica" PR from a while back that does a good job demonstrating the approach. Honestly he's probably better positioned to estimate how much work it would be to dust that PR off and productionize it.

Reply:
> this is a whole different can of worms!

Indeed! But also I think the most scary bit of using a different optimizer just for one particular replica is the potential for mismatch in responses to the controller, and that's in theory mitigated by the controller ignoring all responses for the shadow replica. Definitely would be a lot of replumbing to maintain a different set of dataflow descriptions for just one replica though... potentially so much replumbing as to be inviable. 

> The way I imagined an "Mztrail" would work is that we use the 0dt mechanism to spawn a second read-only environment (a shadow environment!) with a different Mz version and/or different feature flags and run our tests there. This way we could test the whole product (except for sources?), not just the cluster side, and afaict it doesn't require any special infrastructure we don't already get from 0dt. 

For sure—that's how I've imagined it as well. The risk with that approach is that for large environments we may want to test the optimizer on just a single cluster, rather than the entire optimizer. But it might be possible to create a shadow environment that's _limited_ to just a subset of clusters. If you plumbed an "active clusters" set into a read only envd, and taught the coordinator to just drop all other clusters... I think that would Just Work.

Reply:
> Definitely would be a lot of replumbing to maintain a different set of dataflow descriptions for just one replica though... potentially so much replumbing as to be inviable.

Yeah, I'm worried about the amount of complexity that will introduce to the controller. Especially because most of the new code paths will be exercised only when we want to run tests, making it that more likely that they have bugs that bork the environment. It's certainly not impossible to do, but if we're going to implement Mztrail anyway I'd much prefer making that flexible enough to also support our optimizer testing needs!

> But it might be possible to create a shadow environment that's limited to just a subset of clusters.

That seems useful not just for testing the optimizer but also rendering. You can imagine rolling out a new implementation of some operator and wanting to test that on all the production workloads first.

Reply:
Yes, a shadow environment is also how I imagine Mztrail.

> a customer-local record-and-replay operation for handling sources, where we could record spans of customer workloads on customer S3 and replay it on unbilled replicas. Storing it on their S3 means fewer security concerns; record and replay means determinism (and with it, confidence).

The main worry with not doing record and replay is that the data that comes from the sources at the time of running the test can turn out to be not representative of critical workloads? If we do record and replay, we'd still have this issue, but to a lesser extent: we'd need to choose a timeframe to record, and we'd still need to ensure that that timeframe is representative.

> If you plumbed an "active clusters" set into a read only envd, and taught the coordinator to just drop all other clusters...

This would also go well with that idea that comes up from time to time where we'd ask users to mark clusters as "production" clusters. Promising them that we'd test the optimizer on production clusters would give them extra incentive to actually do the marking.

Reply:
The downside to the shadow environment is that it's harder for someone to test with a new optimizer.

Nikhil is not sanguine about customers doing their own qualification, and I suppose we'll ultimately want both things: the ability for us to run shadow environments for mztrail and the ability for customers to experiment with new optimizers. I'm not strongly opinionated on which we do first.

Reply:
I'd be leaning towards us doing the tests. We might need to do tests multiple times (when working on complex fixes), possibly with several customers. So it might get inconvenient to be asking them to do this.

Reply:
Another thing that would help qualification in certain cases is to smarten the upgrade-check tool to make it list all customer plan changes. I don't really know the details of how the upgrade-check tool works, but I imagine this to be a relatively easy thing (compared to, e.g., full mztrail). How much effort do you think this would be, @benesch?

This would help us in cases where a PR will probably not hurt any customer plans, but we are not 100% sure, and skimming through all customer plan changes would either increase our confidence to 100% quickly, or at least inform us about which customers might possibly be regressed, and then we can watch them specifically during the release. Knowing whether a specific customer might possibly be affected by an optimizer change is helpful during a release, because what almost always happens during releases is that at least 1-2 customer clusters start newly OOM looping. There can be various reasons for the OOMs, and it's often hard to figure out the exact reason, and we just somehow end up saying well, it's probably on them: maybe the data size increased, or they were already close to the limit and minor random variations tipped them over, or they installed more dataflows on the cluster, or something. But the problem is that if there was also an optimizer change in the release, then it's harder to confidently say that it's on them.

Reply:
I'm okay with this, certainly to start!

Reply:
> Another thing that would help qualification in certain cases is to smarten the upgrade-check tool to make it list all customer plan changes. I don't really know the details of how the upgrade-check tool works, but I imagine this to be a relatively easy thing (compared to, e.g., full mztrail). How much effort do you think this would be, @benesch?

Relatively easier than a full mztrail, for sure! But still a nontrivial amount of work. I don't think we store the old plans anywhere, so the upgrade check tool would need to somehow extract the plans from the old version in order to be able to compare them to the plans in the new version.

Reply:
Could we run `mzexplore` to export catalogs before and after a 0dt release?

Reply:
For sure, but to what end? If the goal is to have a human review the plans during release qualification, you're better off hooking into the upgrade checker workflow.

Reply:
I should not have said 0dt release, you're right---yes, upgrade checker sounds like the right thing.